internal/client/consensus/grandpa/warp_sync.go

```diff
@@ -0,0 +1,177 @@
+	"fmt"
+
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+const MAX_WARP_SYNC_PROOF_SIZE = 8 * 1024 * 1024
+
+var (
+	errMissingStartBlock      = fmt.Errorf("missing start block")
+	errStartBlockNotFinalized = fmt.Errorf("start block is not finalized")
+)
+
+type BlockState interface {
+	GetHeader(hash common.Hash) (*types.Header, error)
+	BestBlockHeader() (*types.Header, error)
+	GetHeaderByNumber(num uint) (*types.Header, error)
+	GetJustification(hash common.Hash) ([]byte, error)
+}
+
+type GrandpaState interface {
+	GetAuthoritesChangesFromBlock(blockNumber uint) ([]uint, error)
+}
+
+type WarpSyncFragment struct {
+	// The last block that the given authority set finalized. This block should contain a digest
+	// signalling an authority set change from which we can fetch the next authority set.
+	Header types.Header
+	// A justification for the header above which proves its finality. In order to validate it the
+	// verifier must be aware of the authorities and set id for which the justification refers to.
+	Justification GrandpaJustification[hash.H256, uint64]
+}
+
+type WarpSyncProof struct {
+	Proofs []WarpSyncFragment
+	// indicates whether the warp sync has been completed
+	IsFinished   bool
+	proofsLength int
+}
+
+func NewWarpSyncProof() *WarpSyncProof {
+	return &WarpSyncProof{
+		Proofs:       make([]WarpSyncFragment, 0),
+		IsFinished:   false,
+		proofsLength: 0,
+	}
+}
+
+func (w *WarpSyncProof) encode() ([]byte, error) {
+	if w == nil {
+		return nil, fmt.Errorf("cannot encode nil WarpSyncProof")
+	}
+	return scale.Marshal(*w)
+}
+
+func (w *WarpSyncProof) addFragment(fragment WarpSyncFragment) (limitReached bool, err error) {
+	encodedFragment, err := scale.Marshal(fragment)
+	if err != nil {
+		return false, err
+	}
+
+	if w.proofsLength+len(encodedFragment) >= MAX_WARP_SYNC_PROOF_SIZE {
+		return true, nil
+	}
+
+	w.proofsLength += len(encodedFragment)
+	w.Proofs = append(w.Proofs, fragment)
+
+	return false, nil
+}
+
+func (w *WarpSyncProof) lastProofBlockNumber() uint64 {
+	return w.Proofs[len(w.Proofs)-1].Justification.Justification.Commit.TargetNumber + 1
+}
+
+type WarpSyncProofProvider struct {
+	blockState   BlockState
+	grandpaState GrandpaState
+}
+
+// Generate build a warp sync encoded proof starting from the given block hash
+func (np *WarpSyncProofProvider) Generate(start common.Hash) ([]byte, error) {
+	// Get and traverse all GRANDPA authorities changes from the given block hash
+	beginBlockHeader, err := np.blockState.GetHeader(start)
+	if err != nil {
+		return nil, errMissingStartBlock
+	}
+
+	bestLastBlockHeader, err := np.blockState.BestBlockHeader()
+	if err != nil {
+		return nil, err
+	}
+
+	if beginBlockHeader.Number > bestLastBlockHeader.Number {
+		return nil, errStartBlockNotFinalized
```

Comment:
since the error is `errStartBlockNotFinalized` should we use `HighestFinalizedHeader()` instead of  `BestBlockHeader()`?
```suggestion
		return nil, fmt.Errorf("%w: begin block %d, best block %d", errStartBlockNotFinalized, beginBlockHeader.Number, bestLastBlockHeader.Number)
```

Reply:
Now that you are mentioning that, what's the difference between them? I thought the were the same 😅

Reply:
Best block is the tip of the longest chain of unfinalized blocks, i.e. the most recent block produced by BABE. It's the most likely candidate for highest finalized block in the near future.

FWIW, without knowing the details of how warp sync works, I also think we should only consider finalized blocks here.

Reply:
The `BestBlockHeader` is the head of our chain, but it has no requirement to be finalized yet. It can be a recently produced block that grandpa hasn't got to yet. So using `highestFinalizedHeader()` instead will give the highest block that is finalized

Reply:
Got it! thanks for clarifying this! I'll change it 🙌